h2o-automl/src/main/java/ai/h2o/automl/ModelingStep.java

```diff
@@ -316,40 +377,188 @@ protected Work getAllocatedWork() {
+            RandomDiscreteValueSearchCriteria searchCriteria = (RandomDiscreteValueSearchCriteria)buildSpec.build_control.stopping_criteria.getSearchCriteria().clone();
+
             Work work = getAllocatedWork();
+            // for time limit, this is allocated in proportion of the entire work budget.
             double maxAssignedTimeSecs = aml().timeRemainingMs() * getWorkAllocations().remainingWorkRatio(work) / 1e3;
-            // predicate can be removed if/when we decide to include SEs in the max_models limit
-            int maxAssignedModels = (int) Math.ceil(aml().remainingModels() * getWorkAllocations().remainingWorkRatio(work, w -> w._algo != Algo.StackedEnsemble));
+            // SE predicate can be removed if/when we decide to include SEs in the max_models limit
+            // for models limit, this is not assigned in the same proportion as for time,
+            // as the exploitation phase is not supposed to "add" models but just to replace some by better ones,
+            // instead, allocation is done in proportion of the entire exploration budget.
+            int maxAssignedModels = (int) Math.ceil(aml().remainingModels() * getWorkAllocations().remainingWorkRatio(work, isExplorationWork.and(w -> w._algo != Algo.StackedEnsemble)));
 
-            if (searchCriteria.max_runtime_secs() == 0)
-                searchCriteria.set_max_runtime_secs(maxAssignedTimeSecs);
-            else
-                searchCriteria.set_max_runtime_secs(Math.min(searchCriteria.max_runtime_secs(), maxAssignedTimeSecs));
+            searchCriteria.set_max_runtime_secs(searchCriteria.max_runtime_secs() == 0
+                    ? maxAssignedTimeSecs
+                    : Math.min(searchCriteria.max_runtime_secs(), maxAssignedTimeSecs));
 
-            if (searchCriteria.max_models() == 0)
-                searchCriteria.set_max_models(maxAssignedModels);
-            else
-                searchCriteria.set_max_models(Math.min(searchCriteria.max_models(), maxAssignedModels));
+            searchCriteria.set_max_models(searchCriteria.max_models() == 0
+                    ? maxAssignedModels
+                    : Math.min(searchCriteria.max_models(), maxAssignedModels));
 
             if (null == key) key = makeKey(_algo.name(), true);
-            aml().addGridKey(key);
+            aml().trackKey(key);
 
-            aml().eventLog().info(Stage.ModelTraining, "AutoML: starting "+key+" hyperparameter search");
             Log.debug("Hyperparameter search: "+_algo.name()+", time remaining (ms): "+aml().timeRemainingMs());
-            return GridSearch.startGridSearch(
+            return startSearch(
                     key,
                     baseParms,
                     searchParms,
-                    new GridSearch.SimpleParametersBuilderFactory<>(),
-                    searchCriteria,
-                    GridSearch.SEQUENTIAL_MODEL_BUILDING
+                    searchCriteria
             );
         }
     }
+
```

Comment:
Created the relatively complicated abstraction below for 2 reasons:
- it allows to have the exploitation steps for each algo focusing only on the exploitation logic, ie. define the model/search parameters.
- it is generic enough to be used for iterative exploration+selection (e.g. Bayesian Opt, although there are probably small changes needed).